src/main/java/ch/heigvd/pro/b04/questions/QuestionController.java

```diff
@@ -47,58 +51,89 @@ public QuestionController(QuestionRepository repository,
    * @param token The token to check for.
+   * @return Optional with the right moderator, empty if not found
    */
-  private boolean hasAccessAsModerator(int idMod, String token) {
+  private Optional<Moderator> findVerifiedModeratorByIdAndToken(int idMod, String token) {
     return moderatorRepository.findByToken(token)
-        .filter(moderator -> moderator.getIdModerator() == idMod)
-        .isPresent();
+        .filter(moderator -> moderator.getIdModerator() == idMod);
   }
 
   /**
-   * Returns a boolean indicating if the provided token is valid for the participant.
+   * Returns a participant according its id and token. The 2 have to match.
    *
    * @param idMod  The id of the moderator we check access to.
-   * @param idPoll The id of the poll we want to check access to.
    * @param token  The token to check for.
+   * @param idPoll The id of the poll we want to check access to.
+   * @return Optional with the right participant, empty if not found
    */
-  private boolean hasAccessAsParticipant(int idMod, int idPoll, String token) {
+  private Optional<ServerPoll> findVerifiedParticipantByIdAndToken(int idMod, String token,
+      int idPoll) {
     return participantRepository.findByToken(token)
         .map(participant -> participant.getIdParticipant().getIdxServerSession())
         .filter(serverSession -> !(serverSession.getState().equals(SessionState.CLOSED)))
         .map(serverSession -> serverSession.getIdSession().getIdxPoll())
         .filter(poll -> poll.getIdPoll().getIdPoll() == idPoll)
-        .filter(poll -> poll.getIdPoll().getIdxModerator().getIdModerator() == idMod)
-        .isPresent();
+        .filter(poll -> poll.getIdPoll().getIdxModerator().getIdModerator() == idMod);
+  }
+
+  /**
+   * fetch a {@link ServerQuestion} and verify its access by doing it.
+   *
+   * @param idModerator id of the moderator
+   * @param token       token of the moderator
+   * @param idPoll      id of the poll owning the question
+   * @param idQuestion  id of the question to retrieve
+   * @return {@link ServerQuestion} found
+   * @throws WrongCredentialsException if token is not a valid moderator token
+   */
+  private Optional<ServerQuestion> findQuestionByPollAndModerator(int idModerator, String token,
+      int idPoll, int idQuestion) throws WrongCredentialsException {
+    Optional<Moderator> modo = findVerifiedModeratorByIdAndToken(idModerator, token);
+    if (modo.isEmpty()) {
+      throw new WrongCredentialsException();
+    }
+    Optional<ServerQuestion> question = Optional.empty();
+    Optional<ServerPoll> poll = pollRepository.findById(ServerPollIdentifier.builder()
+        .idxModerator(modo.get())
+        .idPoll(idPoll)
+        .build());
+    if (poll.isEmpty()) {
+      return question;
+    }
+
+    return (repository.findById(ServerQuestionIdentifier.builder()
+        .idServerQuestion(idQuestion).idxPoll(poll.get()).build()));
```

Comment:
Use an `Optional<T>.map()` here instead

Reply:
Why is it better ?

Reply:
`poll.get()` is never safer than using a `map`. You want to preserve the functor structure of the `Optional` as much as possible.